### ISSUE TITLE
Allow host key rotation, implement hostkeys-00@openssh.com / hostkeys-prove@openssh.com request

### ISSUE THREAD

**Issue by matthijskooijman** [ronf/asyncssh#667]
Since 2015, OpenSSH has supported returning multiple host keys from the server, to allow a server to operate with an old and new set of keys at the same time to give clients time to update their local known_hosts file. OpenSSH client, through the `UpdateHostKeys` config directive supports updating known_hosts with additional keys supplied by the server (on the condition that one of the supplied keys is already known).

See:
 - [LWN announcement of this release](https://lwn.net/Articles/637156/)
 - [Blogpost about this feature](https://blog.djm.net.au/2015/02/key-rotation-in-openssh-68.html)
 - [OpenSSH protocol extension documentation](https://github.com/openssh/openssh-portable/blob/master/PROTOCOL#L367)

It would be nice if AsyncSSH could support this as well. I'm mostly interested in the server side of this, but it would be good to implement both sides of course.

As for the server side, I tried just passing my old and new key through the `ssh_host_keys` option (it is a list after all), but that gave me: `Error starting server: Multiple keys of type rsa-sha2-256 found`. Without checking extensively, I can imagine the server-side API for this feature could be to lift this restriction of not having duplicate types in the host keys list. The server probably also needs to mark one set of keys as the current/primary ones (which are used in the main connection setup, I suppose and for clients that doe not support these extra requests). This can probably just be "the last (or first) key of each type" or something like that (I think OpenSSH also does this, though docs for the `HostKey` directive are not explicit about this).

**Comment by ronf** [comment]
I've considered this feature, but so far AsyncSSH has strongly adhered to the rule that it never modifies external config files. It will read data from wherever you tell it to load from, and it will default to the same default filenames as OpenSSH to make it easier to share that config, but it NEVER modifies the contents of these files, and this applies both when acting as a client and when acting as a server. It leaves any updates to the config to either the application using AsyncSSH, or it can also take advantage of updates made by OpenSSH with a feature like this when it ends up using the OpenSSH config.

For this host key feature in particular, it seems like there are some ugly edge cases that one would have to deal with. For example, which entries (if any) do you update in the case where the known_hosts entry which matched was based on an SSH certificate and not a key specific to the destination host. Similarly, what get updated in the case where the known_hosts entry matched on some kind of wildcard hostname or IP address pattern? Do you end up adding new entries that are specific to the hostname or IP address the application requests you to connect to even when that isn't the known_hosts entry you matched when making the connection, or do you disable the feature in that case?

Similarly, when is it ok to delete known_hosts entries for keys which no longer show up in the returned list? If an existing known hosts entry refers to a key which no longer appears in the returned list from a host but the entry has multiple patterns and not just a single host and port, should the shared entry be edited to remove just that host/port, or should such shared entries always be left alone?

Would you disable this feature if the application passed it an explicit list of trusted host entries instead of a filename, or used the feature where it can provide a custom callback to validate known hosts? What about the case where an alternate filename is provided that isn't ~/.ssh/known_hosts? Should it update that alternate file if it has write access to it? What do you do when there are multiple known hosts files passed in?

In the end, I'm not sure the value gained from this is worth the development time required and the potential risk of getting something wrong here, especially on the client side.

That said, I think there might be some changes related to this which could be worthwhile and might be more clearly defined.

First, checking out the OpenSSH behavior of the HostKey directive in sshd_config, it looks like it will only ever use the first key of each type as the selected host key in the SSH handshake, but it will list all of the keys when sending the response back to the "hostkeys-00" request. So, if someone were to try and have AsyncSSH use an existing OpenSSH sshd_config that leveraged this feature, AsyncSSH would return an error, which is not good. At least in that case, I'd probably want to ignore the duplicates, even if I didn't support the hostkeys feature. I think I'd still want to return an error when you provide an explicit list with duplicates in it, though, unless AsyncSSH actually added support to return these additional entries in its own "hostkeys-00" server response.

Fully implementing just the server side of this feature in AsyncSSH actually wouldn't be all that difficult, basically just returning the full list of the keys in server_host_keys and then supporting the "prove" feature for each of those keys if I client came back with such a request. This doesn't require modifying any config files, and doesn't have all the edge cases related to trying to decide which known_hosts entries to add/modify/remove. Doing this would allow key rotation to work with OpenSSH clients, and there might be some value in that. There's still a question here about how OpenSSH handles this request if any of its keys have a HostKeyCertificate associated with them, though. In a quick test, it looks like maybe the client requests the keys and server still sends all of them (not clear yet if it is sending the public key or the full cert), but the client actually doesn't seem to do any updating of its local known_hosts when certificates are in use. I'll need to look at this more closely.

If I were to do anything on the client side, I'm thinking maybe the thing to do here is to provide a "hostkeys_handler" argument the application can pass in. This would trigger the client request when it was set, and get back a list of keys from the server. However, the client is supposed to prune the list it gets back for host keys it already trusts and only send "new" keys in the "prove" message, and I'm not sure how to do that. With the existing "known hosts" support on AsyncSSH, there's no way to get a list of all keys that are trusted, and with something like the custom known_hosts callback, that's not even possible. This is getting into the weeds again with the edge cases I listed above. The only option would be to ask the server to "prove" all of the host keys it sends on every request, which is NOT a good idea and could really slow things down.

**Comment by matthijskooijman** [comment]
Thanks for your super-detailed response!

I can see your reservations about the client-side implementation. I agree that using a callback would make sense. As for not requesting proof on all keys: Maybe a callback could receive all host keys, but without requesting the proof automatically, and then offer an API for that callback to request the proof explicitly? Then the application can implement whatever behavior it needs?

As for the server implementation: Good to hear that you can see this implemented without complications, so maybe that would be a good starting point?

Having said that: I do not have any bandwidth for helping with the implementation, and also do not have a specific need for it right now (I wanted to use it to recover from an accidentally changed host key, but I fixed that manually now), I just noticed it was not supported and wanted to create an issue to at least track this :-)

**Comment by ronf** [comment]
Thinking about this a bit more, OpenSSH doesn't actually send the hostkeys-00 request until after authentication fully completes. In fact, in the typical case, it will send this request  in parallel the initial "open channel" request, to avoid adding latency there. So, I probably wouldn't even need to add a callback here. The application could simply call a new API on the SSHClientConnection object if it wanted to get an updated list of host keys, triggering the hostkeys-00 request.

At that point, the client connection object also has a list of all currently trusted host keys, as determined by matching against the known hosts file(s). As it turns out, this works even in the case of a custom known hosts callback, as that is required to return all possible trusted keys up front so that AsyncSSH knows what host key algs to send in the handshake. So, I could potentially use that to decide which host keys to put in a "prove" request. The existing known hosts matching also returns a list of revoked host keys, so those could potentially be excluded from the list of host keys returned, without asking the server to "prove" it has those private keys.

I would still avoid any modifications to existing known hosts files in AsyncSSH itself, but I could potentially have this new API return something like a list of added and removed host public keys, and maybe also a list of revoked public keys, though that information would be coming completely from the client based on its existing known_hosts. I could call this something like get_server_host_key_updates(), sitting alongside the existing get_server_host_key() method on SSHClientConnection which returns the server host key negotiated in the SSH handshake.

I'll experiment with this a bit and let you know how it goes.

**Comment by ronf** [comment]
Quick update: I've got both client and server support for this working, and interoperating with OpenSSH! I still need to add unit tests for this, and there are some other small issues I'm still thinking about, but the bulk of the work is done and it wasn't actually all that complicated (a few hundred line patch).

Right now, like OpenSSH, the server side is always enabled (meaning that the server will send a "hostkeys-00@openssh.com" global request at clients unconditionally. I'm tempted to add a config knob for that, though, as I've heard stories of some clients failing to properly ignore global requests which they don't understand.

On the client side, there's a new connection option called "server_host_keys_handler". When that's set and known hosts checking is enabled, AsyncSSH will send an appropriate "hostkeys-prove-00@openssh.com" request if needed to the server, validate the results, and then call the handler with a list of added, removed, and revoked keys, comparing the known_hosts matches against what the server reported as its current host keys.

The prove request is only needed when there are potential "added" keys in the server's list which are not yet trusted in known_hosts. If that request fails or the signature check fails on any of the keys, those keys are left out of the "added" list.

**Comment by ronf** [comment]
My initial version of this is now available in the "develop" branch as commit 050e5c1. Here's a summary:

- To enable this on the server side, add the option `send_server_host_keys=True` when starting the server.
- To enable this on the client side, specify a callback or coroutine via the option `server_host_keys_handler` when starting the client.

The client callback can be either a `callable` or coroutine, and its signature looks something like:

```python
    def server_host_keys_handler(added: List[SSHKey], removed: List[SSHKey],
                                 retained: List[SSHKey], revoked: List[SSHKey]) -> None:
```

Keys which appear in the server's list but not in known_hosts show up as "added". Keys which appear in known_hosts but not in the server's list any more show up as "removed". Keys which appear in both lists show up as "retained". Keys in the server's list which are marked as "revoked" in known_hosts show up as "revoked".

Note that known_host checking must be enabled for the callback to be called. Also, note that this feature only works on direct host keys, and not host certificates. I've even seen an indication that OpenSSH may disable this feature on the client when the server host key negotiated in the SSH handshake was based on an SSH certificate and certificate authority in known_hosts. That makes some sense, since a certificate could be rotated on the server without needing any updates to the client configuration, at least until the certificate authority installed on the client expires.

I've tested this against itself and against OpenSSH (both as a client and as a server). I'm not sure if there are any other implementations which support this yet.

Feedback is welcome!

**Comment by matthijskooijman** [comment]
Thanks! I had a look over the code, and it looks good to me.

One suggestion: Maybe it would be good to still raise an error in the server when multiple host keys for the same alg are passed, but `send_server_host_keys` is not set? If the error message would hint at the `send_server_host_keys` option, this would make this support more discoverable (the first thing I tried was just add multiple keys, before I learnt that a special request and thus implementation was needed).

**Comment by ronf** [comment]
Good idea - the code once again returns the exception when multiple keys of the same type are configured and send_server_host_keys is disabled, with some additional text suggesting it be enabled.

I also added a note to the documentation for this option explaining this.

**Comment by ronf** [comment]
This feature is now available in AsyncSSH 2.16.0.
